services/src/datasets/external/sentinel_s2_l2a_cogs.rs

```diff
@@ -0,0 +1,640 @@
+use geoengine_datatypes::spatial_reference::{SpatialReference, SpatialReferenceAuthority};
+use geoengine_operators::engine::QueryRectangle;
+use geoengine_operators::source::GdalDatasetParameters;
+use geoengine_operators::source::GdalLoadingInfoPart;
+use geoengine_operators::source::GdalLoadingInfoPartIterator;
+use geoengine_operators::{
+    engine::{MetaData, MetaDataProvider, RasterResultDescriptor, VectorResultDescriptor},
+    mock::MockDatasetDataSourceLoadingInfo,
+    source::{GdalLoadingInfo, OgrSourceDataset},
+};
+use log::debug;
+use serde::{Deserialize, Serialize};
+use snafu::ResultExt;
+use std::collections::HashMap;
+use std::path::PathBuf;
+
+#[derive(Clone, Debug, Serialize, Deserialize)]
+#[serde(rename_all = "camelCase")]
+pub struct SentinelS2L2ACogsProviderDefinition {
+    name: String,
+    id: DatasetProviderId,
+    api_url: String,
+}
+
+#[typetag::serde]
+#[async_trait]
+impl DatasetProviderDefinition for SentinelS2L2ACogsProviderDefinition {
+    async fn initialize(
+        self: Box<Self>,
+    ) -> crate::error::Result<Box<dyn crate::datasets::listing::DatasetProvider>> {
+        Ok(Box::new(SentinelS2L2aCogsDataProvider::new(
+            self.id,
+            self.api_url,
+        )))
+    }
+
+    fn type_name(&self) -> String {
+        "SentinelS2L2ACogs".to_owned()
+    }
+
+    fn name(&self) -> String {
+        self.name.clone()
+    }
+
+    fn id(&self) -> DatasetProviderId {
+        self.id
+    }
+}
+
+#[derive(Debug, Clone)]
+pub struct Band {
+    pub name: String,
+    pub no_data_value: Option<f64>,
+    pub data_type: RasterDataType,
+}
+
+impl Band {
+    pub fn new(name: String, no_data_value: Option<f64>, data_type: RasterDataType) -> Self {
+        Self {
+            name,
+            no_data_value,
+            data_type,
+        }
+    }
+}
+
+#[derive(Debug, Clone)]
+pub struct Zone {
+    pub name: String,
+    pub epsg: u32,
+}
+
+impl Zone {
+    pub fn new(name: String, epsg: u32) -> Self {
+        Self { name, epsg }
+    }
+}
+
+#[derive(Debug, Clone)]
+pub struct SentinelMetaData {
+    bands: Vec<Band>,
+    zones: Vec<Zone>,
+}
+
+#[derive(Debug, Clone)]
+pub struct SentinelDataset {
+    band: Band,
+    zone: Zone,
+    listing: DatasetListing,
+}
+
+pub struct SentinelS2L2aCogsDataProvider {
+    #[allow(dead_code)]
+    id: DatasetProviderId,
+    api_url: String,
+    #[allow(dead_code)]
+    meta_data: SentinelMetaData,
+    datasets: HashMap<DatasetId, SentinelDataset>,
+}
+
+impl SentinelS2L2aCogsDataProvider {
+    pub fn new(id: DatasetProviderId, api_url: String) -> Self {
+        let meta_data = Self::load_metadata();
+        Self {
+            id,
+            api_url,
+            datasets: Self::create_datasets(&id, &meta_data),
+            meta_data,
+        }
+    }
+
+    fn load_metadata() -> SentinelMetaData {
+        // TODO: fetch dataset metadata from config or remote
+        // TODO: is there a no data value?
+        SentinelMetaData {
```

Comment:
this will never scale there are like 12 bands and 80 zones.

Reply:
yeah, we would need hierarchical datasets here I guess

Reply:
or a different selection ui

Reply:
we could have a special dialog for Sentinal datasets but I would prefer a generic solution

Reply:
so add a task for it

Reply:
we already have one https://trello.com/c/ebe6iQwz/115-bring-back-datasets-with-more-then-one-layer-band